modules/glusterfs/storagedlinuxglusterfsvolume.c

```diff
@@ -0,0 +1,203 @@
+ * only be accessed using the provided API.
+ */
+struct _StoragedLinuxGlusterFSVolume
+{
+  StoragedGlusterFSVolumeSkeleton parent_instance;
+};
+
+struct _StoragedLinuxGlusterFSVolumeClass
+{
+  StoragedGlusterFSVolumeSkeletonClass parent_class;
+};
+
+static void glusterfs_volume_iface_init (StoragedGlusterFSVolumeIface *iface);
+
+G_DEFINE_TYPE_WITH_CODE (StoragedLinuxGlusterFSVolume, storaged_linux_glusterfs_volume, STORAGED_TYPE_GLUSTERFS_VOLUME_SKELETON,
+                         G_IMPLEMENT_INTERFACE (STORAGED_TYPE_GLUSTERFS_VOLUME, glusterfs_volume_iface_init));
+
+/* ---------------------------------------------------------------------------------------------------- */
+
+static void
+storaged_linux_glusterfs_volume_finalize (GObject *object)
+{
+  if (G_OBJECT_CLASS (storaged_linux_glusterfs_volume_parent_class)->finalize != NULL)
+    G_OBJECT_CLASS (storaged_linux_glusterfs_volume_parent_class)->finalize (object);
+}
+
+static void
+storaged_linux_glusterfs_volume_init (StoragedLinuxGlusterFSVolume *gfs_volume)
+{
+  g_dbus_interface_skeleton_set_flags (G_DBUS_INTERFACE_SKELETON (gfs_volume),
+                                       G_DBUS_INTERFACE_SKELETON_FLAGS_HANDLE_METHOD_INVOCATIONS_IN_THREAD);
+}
+
+static void
+storaged_linux_glusterfs_volume_class_init (StoragedLinuxGlusterFSVolumeClass *klass)
+{
+  GObjectClass *gobject_class;
+
+  gobject_class = G_OBJECT_CLASS (klass);
+  gobject_class->finalize = storaged_linux_glusterfs_volume_finalize;
+}
+
+/**
+ * storaged_linux_glusterfs_volume_new:
+ *
+ * Creates a new #StoragedLinuxGlusterFSVolume instance.
+ *
+ * Returns: A new #StoragedLinuxGlusterFSVolume. Free with g_object_unref().
+ */
+StoragedGlusterFSVolume *
+storaged_linux_glusterfs_volume_new (void)
+{
+  return STORAGED_GLUSTERFS_VOLUME (g_object_new (STORAGED_TYPE_LINUX_GLUSTERFS_VOLUME,
+                                                  NULL));
+}
+
+static void
+storaged_linux_glusterfs_add_brick_to_volume (StoragedLinuxGlusterFSVolume *gfs_volume,
+                                              GVariant *                   *brick_info)
+
+{
+  const gchar *const *obj_paths;
+  const gchar *brick_name;
+  GString *brick_obj_path;
+  const gchar **p;
+  guint n;
+
+  brick_obj_path = g_string_new ("/org/storaged/Storaged/glusterfs/brick/");
+  obj_paths = storaged_glusterfs_volume_get_bricks (STORAGED_GLUSTERFS_VOLUME (gfs_volume));
+
+  if (g_variant_lookup (brick_info, "name", "&s", &brick_name)) {
+    StoragedLinuxGlusterFSBrickObject *brick_obj;
+
+    for (n = 0; obj_paths != NULL && obj_paths[n] != NULL; n++);
+    p = g_new0 (const gchar *, n + 2);
+
+    storaged_safe_append_to_object_path (brick_obj_path, brick_name);
+    n = 0;
+    while (obj_paths != NULL && obj_paths[n] != NULL) {
+      if (g_strcmp0 (brick_obj_path->str, obj_paths[n]) == 0)
+        return;
+      p[n] = obj_paths[n];
+      n++;
+    }
+    p[n] = brick_obj_path->str;
```

Comment:
Please, use `p[n] = g_string_free (brick_obj_path, FALSE)`. This frees the GString structure without freeing the `->str` member, which is returned instead.